src/server/saved_objects/migrations/core/build_active_mappings.ts

```diff
@@ -38,14 +39,90 @@ export function buildActiveMappings({
   properties: MappingProperties;
 }): IndexMapping {
   const mapping = defaultMapping();
+
+  properties = validateAndMerge(mapping.doc.properties, properties);
+
   return _.cloneDeep({
     doc: {
       ...mapping.doc,
-      properties: validateAndMerge(mapping.doc.properties, properties),
+      properties,
+      _meta: {
+        migrationMappingPropertyHashes: md5Values(properties),
+      },
     },
   });
 }
 
+/**
+ * Diffs the actual vs expected mappings. The properties are compared using md5 hashes stored in _meta, because
+ * actual and expected mappings *can* differ, but if the md5 hashes stored in actual.doc._meta.migrationMappingPropertyHashes
+ * match our expectations, we don't require a migration. This allows ES to tack on additional mappings that Kibana
+ * doesn't know about or expect, without triggering continual migrations.
+ */
+export function diffMappings(actual: IndexMapping, expected: IndexMapping) {
+  if (actual.doc.dynamic !== expected.doc.dynamic) {
+    return { changedProp: 'doc.dynamic' };
+  }
+
+  if (!actual.doc._meta || !actual.doc._meta.migrationMappingPropertyHashes) {
+    return { changedProp: 'doc._meta' };
+  }
+
+  const changedProp = findChangedProp(
+    actual.doc._meta.migrationMappingPropertyHashes,
+    expected.doc._meta!.migrationMappingPropertyHashes
+  );
+
+  return changedProp ? { changedProp: `doc.properties.${changedProp}` } : undefined;
+}
+
+// Convert an object to an md5 hash string, using rison so we get a stable serialization
+function md5Object(obj: any) {
+  return crypto
+    .createHash('md5')
+    .update(canonicalStringify(obj))
+    .digest('hex');
+}
+
+// JSON.stringify is non-canonical, meaning the same object may produce slightly
+// different JSON, depending on compiler optimizations (e.g. object keys
+// are not guaranteed to be sorted). This function consistently produces the same
+// string, if passed an object of the same shape.
+function canonicalStringify(obj: any): string {
+  if (Array.isArray(obj)) {
+    return `[${obj.map(canonicalStringify)}]`;
+  }
+
+  if (!obj || typeof obj !== 'object') {
+    return JSON.stringify(obj);
+  }
+
+  const keys = Object.keys(obj);
+
+  // This is important for properly handling Date
+  if (!keys.length) {
+    return JSON.stringify(obj);
+  }
+
+  const sortedObj = keys
+    .sort((a, b) => a.localeCompare(b))
```

Comment:
Do we really need to use `localeCompare`? I don't think we really care about what order the keys are in, just that they are in the same order every time. If we do need to use this, can you add a comment explaining why?

Reply:
Just that they are in a consistent order is sufficient, so we need to sort in a predictable way. `localeCompare` is just a convenience.